pex/resolver.py

```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
   return Platform.create(platform) if platform and platform != 'current' else None
 
 
+class AtomicDirectory(namedtuple('AtomicDirectory', ['work_dir', 'target_dir'])):
+  @classmethod
+  def for_target_dir(cls, target_dir):
+    return cls(work_dir='{}.{}'.format(target_dir, uuid4().hex), target_dir=target_dir)
+
+  @property
+  def is_finalized(self):
+    return os.path.exists(self.target_dir)
+
+  def finalize(self):
+    if self.is_finalized:
+      return
+
+    try:
+      os.rename(self.work_dir, self.target_dir)
+    except OSError as e:
+      if e.errno != errno.ENOTEMPTY:
+        raise e
+
+
+class ResolveResult(namedtuple('ResolveResult', ['target', 'download_dir'])):
+  @staticmethod
+  def _is_wheel(path):
+    return os.path.isfile(path) and path.endswith('.whl')
+
+  def _iter_distribution_paths(self):
+    if os.path.exists(self.download_dir):
```

Comment:
If `self.download_dir` does not exist, then this will never yield anything, meaning `build_requests` and `install_requests` will also never yield anything. 

Is there a use case where we expect `download_dir` to not exist? If that is not supposed to be supported, then we should raise an exception.

If we do expect it to exist, you can draw more attention to this edge case by inverting the conditional:

```python
def _iter_distribution_paths(self):
  if not os.path.exists(self.download_dir):
    return
  for distribution in ...
    yield ...
```

Reply:
Download dir will not exist for one common case, building a pex from a local project with no `install_requires` and one strange but historically important case, running `pex` (running the pex tool with no args at all which builds a full temporary pex with no requirements and drops you into an interpreter isolted from all site-packages). 

Early return it is - but our tastes here do diverge.